### ISSUE TITLE
visualize/visualize_freeform giving a 302 error (labs)

### ISSUE THREAD

**Issue by cmungall** [geneontology/amigo#184]
I'm trying to paste the output of http://geoffrey.crbs.ucsd.edu:9000/scigraph/graph/neighbors/UBERON:0001831?depth=1&blankNodes=false&direction=both

into
http://amigo2.berkeleybop.org/amigo/visualize_freeform

```
Found

The server encountered an internal error or misconfiguration and was unable to complete your request.

Please contact the server administrator, webmaster@localhost and inform them of the time the error occurred, and anything you might have done that may have caused the error.

More information about this error may be available in the server error log.```

**Comment by kltm** [comment]
Hrm. No immediate obvious answer. Various subgraphs of the given example work, but not the whole. The differences between working and non-working subsets doesn't make any sense yet. Odd.

**Comment by kltm** [comment]
In case we lose the test data at the URL:

```json
{

    "nodes": 

[

{

    "id": "UBERON:0001065",
    "lbl": "parotid duct",
    "meta": 

{

    "category": 

        [
            "anatomical entity"
        ]
    }

},
{

    "id": "200px-Illu_quiz_hn_02.jpg",
    "lbl": null,
    "meta": 

    {
        "category": [ ]
    }

},
{

    "id": "UBERON:0010314",
    "lbl": "structure with developmental contribution from neural crest",
    "meta": 

{

    "category": 

        [
            "anatomical entity"
        ]
    }

},
{

    "id": "UBERON:0003419",
    "lbl": "mesenchyme of parotid",
    "meta": 

{

    "category": 

        [
            "anatomical entity"
        ]
    }

},
{

    "id": "UBERON:0001831PHENOTYPE",
    "lbl": "parotid gland phenotype",
    "meta": 

    {
        "category": [ ]
    }

},
{

    "id": "UBERON:0000409",
    "lbl": "serous gland",
    "meta": 

{

    "category": 

        [
            "anatomical entity"
        ]
    }

},
{

    "id": "UBERON:0004119",
    "lbl": "endoderm-derived structure",
    "meta": 

{

    "category": 

        [
            "anatomical entity"
        ]
    }

},
{

    "id": "Parotid_Gland",
    "lbl": null,
    "meta": 

    {
        "category": [ ]
    }

},
{

    "id": "UBERON:0001567",
    "lbl": "cheek",
    "meta": 

{

    "category": 

        [
            "anatomical entity"
        ]
    }

},
{

    "id": "OBO:uberon/core#efo_slim",
    "lbl": null,
    "meta": 

    {
        "category": [ ]
    }

},
{

    "id": "OBO:uberon/core#uberon_slim",
    "lbl": null,
    "meta": 

    {
        "category": [ ]
    }

},
{

    "id": "Parotid_gland",
    "lbl": null,
    "meta": 

    {
        "category": [ ]
    }

},
{

    "id": "C0030580",
    "lbl": null,
    "meta": 

    {
        "category": [ ]
    }

},
{

    "id": "UBERON:0011847",
    "lbl": "acinus of parotid gland",
    "meta": 

{

    "category": 

        [
            "anatomical entity"
        ]
    }

},
{

    "id": "Illu_quiz_hn_02.jpg",
    "lbl": null,
    "meta": 

    {
        "category": [ ]
    }

},
{

    "id": "UBERON:0001829",
    "lbl": "major salivary gland",
    "meta": 

{

    "category": 

        [
            "anatomical entity"
        ]
    }

},
{

    "id": "181234002",
    "lbl": null,
    "meta": 

    {
        "category": [ ]
    }

},
{

    "id": "UBERON:0003360",
    "lbl": "epithelium of parotid gland",
    "meta": 

{

    "category": 

        [
            "anatomical entity"
        ]
    }

},
{

    "id": "UBERON:0001831",
    "lbl": "parotid gland",
    "meta": 

{

    "category": 

        [
            "anatomical entity"
        ]
    }

},
{

    "id": "UBERON:0012102",
    "lbl": "buccal salivary gland",
    "meta": 

{

    "category": 

        [
            "anatomical entity"
        ]
    }

},
{

    "id": "UBERON:0008801",
    "lbl": "parotid gland primordium",
    "meta": 

{

    "category": 

            [
                "anatomical entity"
            ]
        }
    }

],
"edges":
[

{

    "sub": "UBERON:0001831",
    "obj": "UBERON:0000409",
    "pred": "subClassOf",
    "meta": { }

},
{

    "sub": "UBERON:0001831",
    "obj": "UBERON:0012102",
    "pred": "subClassOf",
    "meta": { }

},
{

    "sub": "UBERON:0001831",
    "obj": "Illu_quiz_hn_02.jpg",
    "pred": "depicted_by",
    "meta": { }

},
{

    "sub": "UBERON:0001831",
    "obj": "181234002",
    "pred": "hasDbXref",
    "meta": { }

},
{

    "sub": "UBERON:0003419",
    "obj": "UBERON:0001831",
    "pred": "BFO_0000050",
    "meta": { }

},
{

    "sub": "UBERON:0001065",
    "obj": "UBERON:0001831",
    "pred": "BFO_0000050",
    "meta": { }

},
{

    "sub": "UBERON:0001831",
    "obj": "UBERON:0001567",
    "pred": "BFO_0000050",
    "meta": { }

},
{

    "sub": "UBERON:0001831",
    "obj": "Parotid_gland",
    "pred": "hasDbXref",
    "meta": { }

},
{

    "sub": "UBERON:0001831PHENOTYPE",
    "obj": "UBERON:0001831",
    "pred": "UPHENO_0000001",
    "meta": { }

},
{

    "sub": "UBERON:0001831",
    "obj": "200px-Illu_quiz_hn_02.jpg",
    "pred": "depicted_by",
    "meta": { }

},
{

    "sub": "UBERON:0011847",
    "obj": "UBERON:0001831",
    "pred": "BFO_0000050",
    "meta": { }

},
{

    "sub": "UBERON:0003360",
    "obj": "UBERON:0001831",
    "pred": "BFO_0000050",
    "meta": { }

},
{

    "sub": "UBERON:0001831",
    "obj": "UBERON:0010314",
    "pred": "subClassOf",
    "meta": { }

},
{

    "sub": "UBERON:0001831",
    "obj": "C0030580",
    "pred": "hasDbXref",
    "meta": { }

},
{

    "sub": "UBERON:0001831",
    "obj": "Parotid_Gland",
    "pred": "hasDbXref",
    "meta": { }

},
{

    "sub": "UBERON:0001831",
    "obj": "UBERON:0008801",
    "pred": "RO_0002202",
    "meta": { }

},
{

    "sub": "UBERON:0001831",
    "obj": "OBO:uberon/core#efo_slim",
    "pred": "inSubset",
    "meta": { }

},
{

    "sub": "UBERON:0001831",
    "obj": "OBO:uberon/core#uberon_slim",
    "pred": "inSubset",
    "meta": { }

},
{

    "sub": "UBERON:0001831",
    "obj": "UBERON:0001829",
    "pred": "subClassOf",
    "meta": { }

},

        {
            "sub": "UBERON:0001831",
            "obj": "UBERON:0004119",
            "pred": "subClassOf",
            "meta": { }
        }
    ]

}
```

**Comment by kltm** [comment]
Started guessing that this might have something to do with input length and the server rather than the client software (also, the CLI attempts seem to mostly work okay). Got an interesting message from the production server that has different settings:

```
 Error message:
malformed header from script. Bad header=0%20%20%22sub%22%3A%20%22UBERO: amigo 
```

Playing around to the point that I can make a set pass or not by playing with labels, which would seem to imply something really weird or server setting kinds of issues.

**Comment by kltm** [comment]
Pretty sure it's not directly AmiGO, CLI of same thing gives:

![cli-01](https://cloud.githubusercontent.com/assets/2687900/7040284/1e3b25d6-dd82-11e4-82c5-26353b2af39c.png)

**Comment by kltm** [comment]
Given that what happens is that the initial POST goes to /amigo/visualize (HTMLClient.pm), which then turns it into a redirect to /visualize (VisualizeServer.pm). Redirects cannot POST, so it is converted into a GET and becomes a URL with length 15404 (measured). The typical server limit is like north of 8k or something, so I'm guessing there is a setting issue.

The "best" fix would be re-working the code to not have redirects. Likely, I'd fold the visualize client into VisualizeServer.pm so there are no redirects, only POSTs. I'm guessing that would be a fix.

Short-term, maybe an apache setting could be found to up the limit sufficiently. A few tries and nothing so far.

Likely quite low priority as it is a labs server setting issue with no current "real" use case. Unless you need this for something @cmungall?

**Comment by cmungall** [comment]
No need for hacky fixes. Testing the viability of driving AmiGO widgets from SciGraph. The final product would be a widget that consumes the bbop/amigo lib and calls SG directly, so the json would come in directly from an api call, not as a get/post param

**Comment by kltm** [comment]
Sorry, I can't do that, Dave.

This /is/ actually an issue, so I'll keep this open until resolved.

This isn't the first time we've had trouble with the redirect mechanism--I thinking that for everything it papers over, it's probably better to avoid it altogether (for example, it took a lot more digging to find the problem, which is likely caused by the system in the first place).

**Comment by kltm** [comment]
Should be fixed with the completion of #185.

**Comment by kltm** [comment]
@cmungall This should now work on labs. Please not that the URLs for visualize components have changed, so access them from the software_list page. A little more work for #185 to be finished, but I'll leave it for now.
